feat: add donut-chart, data-viz-colors and chart-legend

martin-ebay opened the pull request:
<!-- Insert GitHub issue number below -->
Fixes #

<!-- Select which type of PR this is -->
- [x] This PR contains CSS changes
- [ ] This PR does not contain CSS changes

## Description
adding the donut-chart, chart-legend, and data-viz-colors to support the new donut-chart and chart-legend compnents in ebayui-core

## Notes
this is to support a PR that will be opened soon to ebayui-core to add the donut-chart and chart-legend components

## Screenshots
<!-- Upload screenshots of UI before & after these changes -->

## Checklist
<!-- Acknowledge completion of steps in checklists below. Delete lists that are not applicable -->

<!-- For all PR types -->
- [x] I verify the build is in a non-broken state
- [x] I verify all changes are within scope of the linked issue

<!-- For CSS changes -->
- [x] I regenerated all CSS files under dist folder
- [x] I tested the UI in all supported browsers
- [x] I tested the UI in dark mode and RTL mode
- [ ] I added/updated/removed Storybook coverage as appropriate

ianmcburnie commented:
Thanks @martin-ebay . This is going to need a bit of work in order to get it fully BEM compliant. We'll probably need to rename and move a few things round too. Might be quickest if somebody from our side just does that. I'll add a couple of very quick comments now, and we can do a more thorough review later.

ianmcburnie commented on Gemfile.lock:
```diff
@@ -272,4 +272,4 @@ DEPENDENCIES
 
 BUNDLED WITH
-   2.3.18
+   2.3.20
```
Let's remove this change from the PR.

ianmcburnie commented on dist/data-viz-colors/data-viz-colors.css:
```diff
@@ -0,0 +1,85 @@
+.data-viz-color-primary {
+  fill: var(--color-data-viz-primary, var(--color-blue-6));
+}
+.data-viz-color-primary::before {
```
Is every type of data viz going to need these ::before pseudo selectors? Or might it just be specific to the donut chart?

martin-ebay replied:
the ::before is used for the legend bullets

martin-ebay replied:
I'm importing this file into both the chart-legend and donut-chart components in ebayui-core

ianmcburnie commented on dist/donut-chart/donut-chart.css:
```diff
@@ -0,0 +1,18 @@
+  opacity: 1;
+  transition: opacity 0.3s;
+}
+.donut-chart-inactive {
```
Not sure of the use case here, but if this is something that toggles it might be better as a modifier (i.e. .donut-chart--inactive). Alternatively we may consider piggy backing off an ARIA attribute selector if any are suitable.

martin-ebay replied:
good call was meant to be a modifier

ianmcburnie commented on src/less/chart-legend/chart-legend.less:
```diff
@@ -0,0 +1,66 @@
+.chart-legend {
+    display: inline-block;
+}
+.chart-legend .hidden {
```
We prefer to use `[hidden]` rather than `.hidden` if possible

ianmcburnie commented on dist/data-viz-colors/data-viz-colors.css:
```diff
@@ -0,0 +1,85 @@
+  fill: var(--color-data-viz-primary, var(--color-blue-6));
+}
+.data-viz-color-primary::before {
+  background-color: var(--color-data-viz-primary, var(--color-blue-6));
```
We'll need to add these new data-viz token aliases to evo-light.css and evo-dark.css. If there are a LOT of them we might also consider creating separate/optional data-viz token files.

martin-ebay replied:
I added the --color-data-viz-* values just so they could be overwritten if wanted, they use the fallback with the standard --color values

ianmcburnie replied:
First value should be the component level token (i.e. a component level override), second fallback value should be our system token (note: system tokens don't yet exist in skin for data viz, but we should add them)

martin-ebay commented:
@ianmcburnie I've created an evo-data-viz-tokens file, removed the data-viz-colors and moved the styling for the colors into the chart-legend and donut-chart style files as well

ianmcburnie commented on dist/chart-legend/chart-legend.css:
```diff
@@ -0,0 +1,103 @@
+.chart-legend {
+  display: inline-block;
+}
+.chart-legend [hidden] {
```
This seems more like you are trying to clip text for assistive technology rather than hide it completely for all users?

martin-ebay replied:
correct, this is so that a percentage is read out by the screen reader.  Not sure if we want to display the percentage, but I know it was mentioned we should have it for the screen readers.

ianmcburnie replied:
Okay then, in that case we shouldn't use the hidden attribute. As the intention of hidden is to hide for all users.

ianmcburnie commented on src/tokens/evo-data-viz.css:
```diff
@@ -0,0 +1,10 @@
+:root {
+    --color-data-viz-primary: var(--color-blue-6);
```
We need the overrides for dark mode too. Two options:

1. Rename to `evo-data-viz-light.css` and create the corresponding dark file
2. Move these tokens into `evo-light.css` and `evo-dark.css`

Probably easiest to just go with option 2 for now and we can always split them out later if need be.

ianmcburnie replied:
We need the overrides for dark mode too. Two options:

1. Rename to `evo-data-viz-light.css` and create the corresponding dark file
2. Move these tokens into `evo-light.css` and `evo-dark.css`

Probably easiest to just go with option 2 for now and we can always split them out later if need be.

ianmcburnie commented on dist/chart-legend/chart-legend.css:
```diff
@@ -0,0 +1,103 @@
+  position: relative;
+  transition: opacity 0.5s;
+}
+.chart-legend ul li.primary::before {
```
To satisfy BEM, these selectors would need to read something like:

`.chart-legend ul li.chart-legend__item--primary::before`

Which, if you don't need the specificity, can most likely be reduced to:

`.chart-legend__item--primary::before`

ianmcburnie commented on dist/chart-legend/chart-legend.css:
```diff
@@ -0,0 +1,103 @@
+.chart-legend ul li.pattern-2::before {
+  background-color: var(--data-viz-quinary, var(--color-data-viz-quinary));
+}
+.chart-legend.table ul {
```
Similar to the comment above about `primary` and `secondary`, having a classname of `.table` we risk other page CSS cascading and leaking into our component if the page has defined a `.table` class. You might want to consider something like `chart-legend__table`.

ianmcburnie reviewed (changes requested):
Can you add an HTML example section to the docs for each module? That will help us review the HTML structures.

martin-ebay commented:
@ianmcburnie I've made changes based on your feedback, including updating the colors for the dark tokens, and added a stories file for the chart-legend as well.  Let me know if you see any other issues.

ianmcburnie commented on dist/donut-chart/donut-chart.css:
```diff
@@ -0,0 +1,66 @@
+  opacity: 1;
+  transition: opacity 0.3s;
+}
+.donut-chart__segment.primary {
```
Should be `.donut-chart__segment--primary` to be valid BEM.

If we feel that these segment fills might be used outside of donut charts, then we may want to promote them to their own block class with modifier, e.g. 'data-viz-segment data-viz-segment--primary`.

For now though lets go with the first suggestion and we can always refactor later to the second suggestion if we need to.

ianmcburnie commented on src/less/chart-legend/chart-legend.less:
```diff
@@ -0,0 +1,124 @@
+    width: 100%;
+}
+
+.chart-legend .patternfill::before {
```
`.patternfill` is another one we need to either declare as a portable block (i.e. can live in other skin modules) or as a modifier or child of `.chart-legend`

ianmcburnie commented on src/less/chart-legend/stories/chart-legend.stories.js:
```diff
@@ -0,0 +1,31 @@
+`;
+
+export const patternfill = () => `
+<div legend-style="minimal" patternfill="" class="chart-legend">
```
Should these be data attributes? e.g. `data-chart-legend-style` and `data-chart-patternfill`?

martin-ebay replied:
I think I need to add these to the ignore attributes so they don't get added, I had copy pasted from a live running version from ebayui-core and glossed over these being added.

ArtBlue commented on src/less/chart-legend/chart-legend.less:
```diff
@@ -0,0 +1,123 @@
+    line-height: 2rem;
+    position: relative;
+    transition: opacity 0.5s;
+    &--patternfill::before {
```
Even though `less` offers this internal cascading feature, we typically opt to write out the entire selector. This would apply to all of the internal ones as well.

ianmcburnie replied:
Agree with @ArtBlue here. Having the entire selector makes it possible to find with text search. This should be added to our style-guide: https://github.com/eBay/skin/blob/master/CONTRIBUTING.md#style-guide

ArtBlue commented on src/less/chart-legend/chart-legend.less:
```diff
@@ -0,0 +1,123 @@
+    width: 10px;
+}
+.chart-legend ul li::after {
+    background-color: rgba(0, 0, 0, 0);
```
Is this really needed?

martin-ebay replied:
it is not, accidental leftover from testing.

ArtBlue commented:
@ianmcburnie , do we need some documentation on the Skin site to detail usage, at least for the abstracted pieces?

ianmcburnie commented:
@ArtBlue Yes, definitely.

martin-ebay commented:
@ianmcburnie and @ArtBlue I'll get the docs updated today

ArtBlue commented:
Hmm...why can I not pull down the branch? 

`git fetch`

`git co martin-ebay:donut-chart`

`error: pathspec 'martin-ebay:donut-chart' did not match any file(s) known to git`

agliga commented:
> Hmm...why can I not pull down the branch?
> 
> `git fetch`
> 
> `git co martin-ebay:donut-chart`
> 
> `error: pathspec 'martin-ebay:donut-chart' did not match any file(s) known to git`

You need to add his fork using `git remote add martin path-to-fork` (ssh clone path)
Then do a `git fetch —all`
And then checkout using `martin` or whatever name you added the fork with in remote add.

ArtBlue commented:
> You need to add his fork using `git remote add martin path-to-fork` (ssh clone path) Then do a `git fetch —all` And then checkout using `martin` or whatever name you added the fork with in remote add.

Thanks! I didn't realize it was in a different repo. I actually found an easier way; just check out the pr branch. I never knew this existed. Very slick.

`gh pr checkout 1837`

martin-ebay commented:
that's cool, I didn't know you could checkout a PR from a fork like that

agliga reviewed (commented):
Overall I think it's great!

I have a couple of minor comments, but I'm fine with it going in like this.

Couple of questions:
1. Do you think the API will change? If so maybe we can add an `alpha` tag to it to indicate the api might change between minor versions. Or hide it from the doc.
2. How will this change with the high charts code? Or will it stay the same.

martin-ebay commented on src/less/chart-legend/chart-legend.less:
```diff
@@ -0,0 +1,122 @@
+    width: 12px;
+}
+.chart-legend__item--primary::before {
+    background-color: var(--data-viz-primary, var(--color-data-viz-primary));
```
I intentionally set it to --data-viz-primary so that the donut-chart__segment--primary would be updated if it was defined as the legend and donut chart segment should match for accessibility

agliga replied:
So usually we want those first ones to be unique even across components. 
they aren’t used yet but they allow for customization later down the line

martin-ebay commented:
@agliga I do think it will change based on the new colors / patterns in the updated design.  I'm thinking I might shift focus later today and start updating the donut chart to match.

As far as will the code change with Highcharts I wasn't planning on it.  But we could always revisit it after.

agliga commented:
Closing for now as per Martin.

With highcharts this might change so we will revisit later. 

Possibly might create a new repo for charts.
